package com.abhishek.ecommercebackendsystem.ExceptionHandler;

import com.abhishek.ecommercebackendsystem.Dtos.ExceptionDto;
import com.abhishek.ecommercebackendsystem.Exceptions.*;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.ControllerAdvice;
import org.springframework.web.bind.annotation.ExceptionHandler;

@ControllerAdvice
public class GlobalExceptionHandler {
    @ExceptionHandler(ArithmeticException.class)
    public ResponseEntity<ExceptionDto> handleArithmeticException(ArithmeticException ex){
        ExceptionDto exceptionDto = new ExceptionDto();
        exceptionDto.setMessage("Something went wrong");
        exceptionDto.setResolution("Arithmetic Exception");
        ResponseEntity<ExceptionDto> responseEntity = new ResponseEntity<>(exceptionDto, HttpStatus.BAD_REQUEST);
        return responseEntity;
    }

    @ExceptionHandler(ArrayIndexOutOfBoundsException.class)
    public ResponseEntity<ExceptionDto> handleArrayIndexOutOfBoundsException(ArrayIndexOutOfBoundsException ex){
        ExceptionDto exceptionDto = new ExceptionDto();
        exceptionDto.setMessage("Something went wrong");
        exceptionDto.setResolution("ArrayIndexOutOfBoundsException");
        ResponseEntity<ExceptionDto> responseEntity = new ResponseEntity<>(exceptionDto, HttpStatus.BAD_REQUEST);
        return responseEntity;
    }

    @ExceptionHandler(RuntimeException.class)
    public ResponseEntity<ExceptionDto> handleRuntimeException(RuntimeException ex){
        ExceptionDto exceptionDto = new ExceptionDto();
        exceptionDto.setMessage("Something went wrong");
        exceptionDto.setResolution("RuntimeException");
        ResponseEntity<ExceptionDto> responseEntity = new ResponseEntity<>(exceptionDto, HttpStatus.BAD_REQUEST);
        return responseEntity;
    }

    @ExceptionHandler(InvalidProductIdException.class)
    public ResponseEntity<ExceptionDto> handleInvalidProductIdException(InvalidProductIdException invalidProductIdException){
        ExceptionDto exceptionDto = new ExceptionDto();
        exceptionDto.setMessage("Invalid Product Id " + invalidProductIdException.getId() + " passed.");
        exceptionDto.setResolution("InvalidProductIdException");
        ResponseEntity<ExceptionDto> responseEntity = new ResponseEntity<>(exceptionDto, HttpStatus.BAD_REQUEST);
        return responseEntity;
    }
    @ExceptionHandler(InvalidCategoryIdException.class)
    public ResponseEntity<ExceptionDto> handleInvalidCategoryIdException(InvalidCategoryIdException invalidCategoryIdException){
        ExceptionDto exceptionDto = new ExceptionDto();
        exceptionDto.setMessage("Invalid Product Id " + invalidCategoryIdException.getId() + " passed.");
        exceptionDto.setResolution("InvalidCategoryIdException");
        ResponseEntity<ExceptionDto> responseEntity = new ResponseEntity<>(exceptionDto, HttpStatus.BAD_REQUEST);
        return responseEntity;
    }
    @ExceptionHandler(NoProductsFoundException.class)
    public ResponseEntity<ExceptionDto> handleNoProductsFoundException(NoProductsFoundException ex){
        ExceptionDto exceptionDto = new ExceptionDto();
        exceptionDto.setMessage("Something Went Wrong");
        exceptionDto.setResolution("NoProductsFoundException");
        ResponseEntity<ExceptionDto> responseEntity = new ResponseEntity<>(exceptionDto, HttpStatus.BAD_REQUEST);
        return responseEntity;
    }
    @ExceptionHandler(NoCategoriesFoundException.class)
    public ResponseEntity<ExceptionDto> handleNoCategoriesFoundException(NoCategoriesFoundException ex){
        ExceptionDto exceptionDto = new ExceptionDto();
        exceptionDto.setMessage("Something Went Wrong");
        exceptionDto.setResolution("NoCategoriesFoundException");
        ResponseEntity<ExceptionDto> responseEntity = new ResponseEntity<>(exceptionDto, HttpStatus.BAD_REQUEST);
        return responseEntity;
    }

    @ExceptionHandler(InvalidCustomerIdException.class)
    public ResponseEntity<ExceptionDto> handleInvalidCustomerIdException(InvalidCustomerIdException ex){
        ExceptionDto exceptionDto = new ExceptionDto();
        exceptionDto.setMessage("Invalid Customer Id " + ex.getId() + " passed.");
        exceptionDto.setResolution("InvalidCustomerIdException");
        ResponseEntity<ExceptionDto> responseEntity = new ResponseEntity<>(exceptionDto, HttpStatus.BAD_REQUEST);
        return responseEntity;
    }

    @ExceptionHandler(NoCustomerFoundException.class)
    public ResponseEntity<ExceptionDto> handleNoCustomerFoundException(NoCustomerFoundException ex){
        ExceptionDto exceptionDto = new ExceptionDto();
        exceptionDto.setMessage("Something Went Wrong");
        exceptionDto.setResolution("NoCustomerFoundException");
        ResponseEntity<ExceptionDto> responseEntity = new ResponseEntity<>(exceptionDto, HttpStatus.BAD_REQUEST);
        return responseEntity;
    }

    @ExceptionHandler(ProductOutOfStockException.class)
    public ResponseEntity<ExceptionDto> handleProductOutOfStockException(ProductOutOfStockException productOutOfStockException){
        ExceptionDto exceptionDto = new ExceptionDto();
        exceptionDto.setMessage("Something Went Wrong");
        exceptionDto.setResolution("Product With ID " +  productOutOfStockException.getId() + " is currently out of stock.");
        ResponseEntity<ExceptionDto> responseEntity = new ResponseEntity<>(exceptionDto, HttpStatus.BAD_REQUEST);
        return responseEntity;
    }

    @ExceptionHandler(InvalidOrderIdException.class)
    public ResponseEntity<ExceptionDto> handleInvalidOrderIdException(InvalidOrderIdException invalidOrderIdException){
        ExceptionDto exceptionDto = new ExceptionDto();
        exceptionDto.setMessage("Something Went Wrong");
        exceptionDto.setResolution("Order with id " + invalidOrderIdException.getId() + " not found.");
        ResponseEntity<ExceptionDto> responseEntity = new ResponseEntity<>(exceptionDto, HttpStatus.BAD_REQUEST);
        return responseEntity;
    }

    @ExceptionHandler(NoOrderHistoryFoundException.class)
    public ResponseEntity<ExceptionDto> handleNoOrderHistoryFoundException(NoOrderHistoryFoundException ex){
        ExceptionDto exceptionDto = new ExceptionDto();
        exceptionDto.setMessage("Something Went Wrong");
        exceptionDto.setResolution("NoOrderHistoryFoundException");
        ResponseEntity<ExceptionDto> responseEntity = new ResponseEntity<>(exceptionDto, HttpStatus.BAD_REQUEST);
        return responseEntity;
    }

    @ExceptionHandler(Exception.class)
    public ResponseEntity<ExceptionDto> handleGeneralException(Exception ex){
        ExceptionDto exceptionDto = new ExceptionDto();
        exceptionDto.setMessage("Something went wrong");
        exceptionDto.setResolution("GeneralException");
        ResponseEntity<ExceptionDto> responseEntity = new ResponseEntity<>(exceptionDto, HttpStatus.BAD_REQUEST);
        return responseEntity;
    }
}
